### DISCUSSION TITLE
Milvus 2 - get list of ids in a collection

### DISCUSSION THREAD

**Discussion by viv1729** [milvus-io/milvus#15481] [Q&A and General discussion] [answered]
Hi,
I decided to switch from Milvus 1 to Milvus 2 because I wanted to have some filters during search that Milvus 2 offers. 
However, it seems that Milvus doesn't have `list_id_in_segment()` method. Please let me know if there is an alternate way to list all the ids from a collection in Milvus 2 - I couldn't find it in docs.

I my use case, indexing happens from a different pipeline where some entities are indexed or not based on some validation. Thus it's quite useful to know what are the indexed ids in a collection at a given time.

**Comment by xiaofan-luan** [comment]
hi vivek, so your goal is to fetch all the ids in the collection/partiton i'm correct?

**Reply by viv1729** [reply]
Hi,
Yes, you are right - that's the query.

**Comment by yhmo** [comment, marked as answer]
@viv1729 
The milvus 2.0 has a method 'query()' to fetch entities from a collection.
Assume there is a collection named "aaa", it has a field named 'id', assume all id values are greater than 0.
```
collection = Collection("aaa")
result = collection.query(expr="id >= 0")
print(result)
```
The result is a list, you will see all the ids are in this list.

**Reply by viv1729** [reply]
Hi @yhmo,
Thank you for the solution. However, I am observing something different number of results compared to what's shown by `num_entities`. I am wondering if it's because there might be duplicates `id` in collection and query only returns unique results?

```Python
print(f"{collection.num_entities:,}")
print(collection.partitions)
print(collection.schema)

1,107,488
[{"name": "_default", "collection_name": "collection_02_2022_02_02_20_35", "description": ""}]
{
  auto_id: False
  description: vsse
  fields: [{
    name: id
    description: int64
    type: 5
    is_primary: True
    auto_id: False
  }, {
    name: float_vector
    description: float vector
    type: 101
    params: {'dim': 8}
  }]
}
```

However, num of results from `query` is much lower:
```Python
result = collection.query(expr="id >= 0")
print(f"{len(result):,}")

# there are no -ve ids that I am indexing, still just to confirm 
result = collection.query(expr="id >= 0 or id < 0")
print(f"{len(result):,}")

392,356
392,356
```

**Reply by yhmo** [reply]
I guess there are a lot of duplicate id entities in your collection.
Use this script to test:
```
import random
from pymilvus import (
    connections,
    utility,
    FieldSchema, CollectionSchema, DataType,
    Collection,
)

print("start connecting to Milvus")
connections.connect("default", host="localhost", port="19530")

collection_name = "TEST"
if utility.has_collection(collection_name):
    utility.drop_collection(collection_name)

fields = [
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False),
    FieldSchema(name="embeddings", dtype=DataType.FLOAT_VECTOR, dim=8)
]

schema = CollectionSchema(fields, "test")

print("Create collection", collection_name)
collection = Collection(collection_name, schema)

print("Start inserting entities")
num_entities = 10000
for k in range(50):
    print('No.', k)
    entities = [
        # [i for i in range(num_entities)], # duplicate id, the query will get 10000 ids
        [i + num_entities*k for i in range(num_entities)],  # unique id, the query will get 500000 ids
        [[random.random() for _ in range(8)] for _ in range(num_entities)],
    ]
    insert_result = collection.insert(entities)

print(f"Number of entities: {collection.num_entities}")
print("Start loading")
collection.load()

result = collection.query(expr="id >= 0")
print("query result count:", len(result))

```
The two lines will get different results for query:
```
[i for i in range(num_entities)], # duplicate id, the query() will get 10000 ids
[i + num_entities*k for i in range(num_entities)],  # unique id, the query() will get 500000 ids
```

**Reply by xiaofan-luan** [reply]
> 
two possibilities
1. Due to delete?
2. Due to duplicate ids, which should only be possible if you specify id on insertion

**Reply by viv1729** [reply]
Hi @yhmo 
Yes, I am able to verify the result of query() having only unique entities. 

@xiaofan-luan 
Yes, it's coz I am inserting using id field. 

I have a follow up question:

-  Duplication of id is happening coz I am doing incremental indexing i.e. whenever an entity gets updated -> some event triggers -> entity gets indexed. So, to avoid duplication, would you recommend using deletion before inserting?
    `collection.delete(expr=f"id in {ids_to_insert}")` # I insert in batch of >=5000, so `len(ids_to_insert) >= 5000`

- Also, to confirm, if I don't avoid duplication then search/query can return data belonging to any of the duplicate id as mentioned [here](https://milvus.io/docs/v2.0.0/product_faq.md#When-vectors-with-duplicate-primary-keys-are-inserted-does-Milvus-treat-it-as-an-update-operation)?

**Reply by xiaofan-luan** [reply]
1.you probably want to delete then insert for now, and latter on we should support update.
2. yes, if your older vector is closer to the query, and the newest vector is not in topK, then older value might be retuerned

**Reply by viv1729** [reply]
Got it. thank you 🙂
